### PR TITLE
Add GRPC adhoc implementation

### PR DESCRIPTION
**Pull request by Bplotka**
Add proto definition for gRPC adhoc rules
Update comments to reflect that SRV lookup remains unimplemented

Generate protobuf

Add new gRPC adhoc generated protobuf

### CODE REVIEW

File: pkg/kedge/grpc/director/director.go [hunk 1]
```diff
@@ -17,18 +19,30 @@ import (
...
-		beName, err := router.Route(ctx, fullMethodName)
+		beName, err := grpcRouter.Route(ctx, fullMethodName)
 		if err != nil {
+			if err == router.ErrRouteNotFound {
+				fmt.Printf("director.go:27 debugging %#v\n", metautils.ExtractIncoming(ctx).Get(":authority"))
+				ipPort, err := adhocRouter.Address(metautils.ExtractIncoming(ctx).Get(":authority"))
+				if err != nil {
+					return ctx, nil, err
+				}
+				// I don't know which other dial options to include here
```
Comment (Bplotka) [inline pkg/kedge/grpc/director/director.go]: you can inspire from: `kedge/grpc/backendpool/backend.go:125`

```
func BuildClientConn(cnf *pb.Backend, target string, resolver naming.Resolver) (*grpc.ClientConn, error) {
	var opts []grpc.DialOption
	opts = append(opts, chooseDialFuncOpt(cnf))
	opts = append(opts, chooseSecurityOpt(cnf))
	opts = append(opts, grpc.WithCodec(proxy.Codec())) // needed for the director to function at all.
	opts = append(opts, chooseInterceptors(cnf)...)
	opts = append(opts, grpc.WithBalancer(&addrTagBalancer{chooseBalancerPolicy(cnf, resolver)}))
	return grpc.Dial(target, opts...)
}
```
I think yours are fine but yu need also: `grpc.WithCodec(proxy.Codec()))` that might solve your issue

File: pkg/kedge/grpc/director/adhoc/adhoc.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	for _, rule := range a.rules {
+		if !common.HostMatches(hostName, rule.DnsNameMatcher) {
+			continue
+		}
+		portForRule := port
+		if port == 0 {
+			if defPort := rule.Port.Default; defPort != 0 {
+				portForRule = int(defPort)
+			} else {
+				portForRule = 81
```
Comment (adamhosier) [inline pkg/kedge/grpc/director/adhoc/adhoc.go]: why port 81 ooi?

File: pkg/kedge/grpc/director/director.go [hunk 3]
```diff
@@ -8,27 +8,53 @@ import (
...
+				var opts []grpc.DialOption
+				opts = append(opts,
+					grpc.WithInsecure(),
+					grpc.WithBlock(),
+					grpc.WithAuthority(fullMethodName),
+					grpc.WithCodec(proxy.Codec()),
+				)
+				cc, err := grpc.Dial(ipPort, opts...)
+				if err != nil {
+					return ctx, nil, errors.Wrap(err, "dial")
```
Comment (adamhosier) [inline pkg/kedge/grpc/director/director.go]: can you improve the error here?

File: pkg/kedge/grpc/director/director.go [hunk 4]
```diff
@@ -8,27 +8,53 @@ import (
...
 )
 
 // New builds a StreamDirector based off a backend pool and a router.
-func New(pool backendpool.Pool, router router.Router) proxy.StreamDirector {
+func New(pool backendpool.Pool, adhocRouter common.Addresser, grpcRouter router.Router) proxy.StreamDirector {
 	return func(ctx context.Context, fullMethodName string) (context.Context, *grpc.ClientConn, error) {
-		beName, err := router.Route(ctx, fullMethodName)
+		beName, err := grpcRouter.Route(ctx, fullMethodName)
 		if err != nil {
+			if err == router.ErrRouteNotFound {
```
Comment (adamhosier) [inline pkg/kedge/grpc/director/director.go]: can this be changed to a slightly flatter strucuture?
```
if err == router.ErrRouteNotFound { ...  return ... }
if err != nil { return ... }
```

File: pkg/kedge/grpc/director/adhoc/adhoc.go (hunk 2 above)
Comment (apuchitnis) [reply pkg/kedge/grpc/director/adhoc/adhoc.go]: This is the default grpc port

**Comment by apuchitnis** [comment]
New PR: https://github.com/improbable-eng/kedge/pull/129

**Pull request closed without merging**